dion/fron.py

```diff
@@ -0,0 +1,785 @@
+    G: List[Tensor],  # local shards (regular Tensors)
+    STATES: List[dict],  # layer-sharded optimizer states (each has 'momentum_full')
+    lr: Tensor,
+    ef_decay: Tensor,
+    fraction: Tensor,
+    weight_decay: Tensor,
+    epsilon: Tensor,
+    flatten: bool,
+    adjust_lr: Optional[str],
+    device_rank: int,
+    world_size: int,
+    shard_dim: Optional[int] = None,
+    process_group: Optional[ProcessGroup] = None,
+    newton_schulz_func: Optional[Callable] = None,
+) -> Generator[None, None, None]:
+    """
+    FRON layer-sharded path:
+      - Matrix-dim sharded: all_to_all shards <-> full, compute once on owner, all_to_all back.
+      - Replicated/unsharded: compute once on owner (index=device_rank), all_gather dense updates.
+      - Single-GPU (batch size=1): compute once on owner, apply locally.
+    """
+
+    # Ownership-by-index: owner of batch index j is rank j.
+    assert 0 <= device_rank < world_size
+    assert len(X) == len(STATES) == world_size  # Guaranteed by padding
+
+    # convert gradients to local shards
+    G_local = to_local(G)
+    frac = float(fraction)
+
+    # Compute adjusted lr from global shape (matrix dims last)
+    if adjust_lr is None:
+        adjusted_lr = lr
+    elif adjust_lr == "spectral_norm":
+        adjusted_lr = adjust_lr_spectral_norm(lr, X[0].shape, flatten=flatten)
+    elif adjust_lr == "rms_norm":
+        adjusted_lr = adjust_lr_rms_norm(lr, X[0].shape, flatten=flatten)
+    else:
+        raise ValueError(f"Unknown adjust_lr value: {adjust_lr}")
+
+    # Matrix-dimension sharded path
+    if shard_dim is not None:
+        assert len(X) == world_size, "Batch size must equal world size"
+        assert (
+            process_group is not None
+        ), "process_group must be provided for sharded DTensors"
+        assert isinstance(X[0], DTensor), "X should contain DTensors"
+        assert X[0].size(shard_dim) % world_size == 0, (
+            f"Shard dimension {shard_dim} size {X[0].size(shard_dim)} "
+            f"is not divisible by world size {world_size}."
+        )
+
+        # Shards -> single full gradient on owner (bf16 comm)
+        G_bf16 = [g.to(dtype=torch.bfloat16) for g in G_local]
+        recv_shards = [torch.empty_like(g) for g in G_bf16]
+        work = dist.all_to_all(recv_shards, G_bf16, group=process_group, async_op=True)
+        yield
+        work.wait()
+
+        full_grad_bf16 = torch.cat(recv_shards, dim=shard_dim)
+
+        # Ownership-by-index contract:
+        # For layer-sharded path, the "owner" of batch index j is rank j (device_rank).
+        owner_state = STATES[device_rank]
+        owner_is_pad = owner_state.get("is_pad", False)
+
+        if owner_is_pad:
+            # For pads, do NOT allocate momentum_full or run NS.
+            # Just return zero shards.
+            send_shards = [torch.zeros_like(g) for g in G_bf16]  # bf16 payloads
+            U = [torch.empty_like(g) for g in G_bf16]
```

Comment:
These lines are duplicated in both branches of the if statement. I think it makes more sense to deduplicate and move it outside the if-else block. Logically, only `send_shards` is different, and the same collective call is used regardless of `owner_is_pad`